internal/pkg/cli/deploy/deploy.go

```diff
@@ -0,0 +1,838 @@
+	ZipAndUpload(bucket, key string, files ...s3.NamedBinary) (string, error)
+}
+
+// Templater stringifies a golang template.
+type Templater interface {
+	Template() (string, error)
+}
+
+// EndpointGetter gets the service discovery endpoint.
+type EndpointGetter interface {
+	ServiceDiscoveryEndpoint() (string, error)
+}
+
+// VersionGetter gets the version.
+type VersionGetter interface {
+	Version() (string, error)
+}
+
+// PublicCIDRBlocksGetter gets the public CIDR blocks.
+type PublicCIDRBlocksGetter interface {
+	PublicCIDRBlocks() ([]string, error)
+}
+
+// CustomResourcesUploader uploads the custom resource files to S3.
+type CustomResourcesUploader interface {
+	UploadEnvironmentCustomResources(upload s3.CompressAndUploadFunc) (map[string]string, error)
+	UploadRequestDrivenWebServiceCustomResources(upload s3.CompressAndUploadFunc) (map[string]string, error)
+	UploadNetworkLoadBalancedWebServiceCustomResources(upload s3.CompressAndUploadFunc) (map[string]string, error)
+}
+
+// SNSTopicsLister lists SNS topics.
+type SNSTopicsLister interface {
+	ListSNSTopics(appName string, envName string) ([]deploy.Topic, error)
+}
+
+// ServiceDeployer uses CloudFormation to deploy a service.
+type ServiceDeployer interface {
+	DeployService(out progress.FileWriter, conf cloudformation.StackConfiguration, bucketName string, opts ...awscloudformation.StackOption) error
+}
+
+// ServiceForceUpdater force updates a service.
+type ServiceForceUpdater interface {
+	ForceUpdateService(app, env, svc string) error
+	LastUpdatedAt(app, env, svc string) (time.Time, error)
+}
+
+// WorkspaceReader reads from the workspace.
+type WorkspaceReader interface {
+	manifestReader
+	pathFinder
+}
+
+// Interpolator interpolates a manifest.
+type Interpolator interface {
+	Interpolate(s string) (string, error)
+}
+
+// progress is the interface to inform the user that a long operation is taking place.
+type Progress interface {
+	Start(label string)
+	Stop(label string)
+}
+
+type manifestReader interface {
+	ReadWorkloadManifest(name string) (workspace.WorkloadManifest, error)
+}
+
+type pathFinder interface {
+	Path() (string, error)
+}
+
+type fileReader interface {
+	ReadFile(string) ([]byte, error)
+}
+
+type workloadDeployer struct {
+	name          string
+	app           *config.Application
+	env           *config.Environment
+	imageTag      string
+	s3Bucket      string
+	mft           interface{}
+	workspacePath string
+}
+
+// WorkloadDeployerInput is the input to for workloadDeployer constructor.
+type WorkloadDeployerInput struct {
+	Name     string
+	App      *config.Application
+	Env      *config.Environment
+	ImageTag string
+	S3Bucket string
+}
+
+// NewWorkloadDeployer is the constructor for workloadDeployer.
+func NewWorkloadDeployer(in *WorkloadDeployerInput) (*workloadDeployer, error) {
+	ws, err := workspace.New()
+	if err != nil {
+		return nil, fmt.Errorf("new workspace: %w", err)
+	}
+	mft, err := workloadManifest(&workloadManifestInput{
```

Comment:
Sorry for being so late to the PR!!! 🙇🏼‍♀️ Just a concern to throw out here: I wonder if it should be considered a side effect that we are actually reading/applying env override/validating the manifest here, in a constructor 🤔 I feel like constructor, ideally should only init a few members, without doing the extra work mentioned just now.

Reply:
The reason why I put the manifest reading in the constructor is because it is used in both `UploadArtifacts` and `DeployWorkload` (`GenerateCloudFormationTemplate` for `workload package` in future PR as well). If we don't handle the logic in the constructor, then we have to either 1) define it as a function within the `cli` package and call it for `workload deploy/package` or 2) call `workloadManifest` within each of the `workloadDeployer` method. What do you think?

Reply:
Ya I like the first option! I feel like it makes sense to not delegate the manifest responsibility to `deploy` package, since manifest operations are not unique to deployment - it is also needed for packaging (is this correct?)

Reply:
> I feel like it makes sense to not delegate the manifest responsibility to deploy package

That makes sense to me. I'll make the change in the new PR.

> since manifest operations are not unique to deployment - it is also needed for packaging (is this correct?)

yes it is required but the thing is i put here also because it is used in `packaging` since as I mentioned `GenerateCloudFormationTemplate` also needs the applied manifest.